File: src/server/runtime.ts
import { homedir } from "node:os";
import { join } from "node:path";
import { AuthService } from "../main/auth/auth-service.js";
import { AuthSessionStore } from "../main/auth/session-store.js";
import { ProviderFacade } from "../main/network/facade.js";
import { CodexProviderAdapter } from "../main/network/providers/codex-provider.js";
import { GeminiProviderAdapter } from "../main/network/providers/gemini-provider.js";
import { OpenAiProviderAdapter } from "../main/network/providers/openai-provider.js";
import type { ProviderRuntimeConfig } from "../main/network/provider-registry.js";
import { DEFAULT_BRIDGE_MODEL } from "../shared/bridge.js";
import type { BridgeServerConfig } from "./types.js";

export type BridgeRuntime = {
  authService: AuthService;
  providerFacade: ProviderFacade;
};

export async function createBridgeRuntime(config?: BridgeServerConfig): Promise<BridgeRuntime> {
  const authStorePath =
    config?.authStorePath ?? join(homedir(), ".codex-bridge", "auth", "codex-session.json");

  const runtimeConfig: ProviderRuntimeConfig = {
    codex: config?.codexBaseUrl
      ? {
          baseUrl: config.codexBaseUrl,
          userAgent: `codex-bridge/${config?.model?.trim() || DEFAULT_BRIDGE_MODEL}`
        }
      : {
          userAgent: `codex-bridge/${config?.model?.trim() || DEFAULT_BRIDGE_MODEL}`
        },
    openai:
      config?.openaiApiKey || config?.openaiBaseUrl
        ? {
            apiKey: config.openaiApiKey,
            baseUrl: config.openaiBaseUrl
          }
        : undefined,
    gemini:
      config?.geminiApiKey || config?.geminiBaseUrl
        ? {
            apiKey: config.geminiApiKey,
            baseUrl: config.geminiBaseUrl
          }
        : undefined
  };

  const authService = new AuthService({
    sessionStore: new AuthSessionStore(authStorePath)
  });
  await authService.initialize();

  const providerFacade = new ProviderFacade({
    authService,
    config: runtimeConfig,
    adapters: [new CodexProviderAdapter(), new OpenAiProviderAdapter(), new GeminiProviderAdapter()]
  });

  return { authService, providerFacade };
}

File: broker/src/codex_bridge_broker/runtime.py
from __future__ import annotations

from dataclasses import dataclass

from .auth import AuthService
from .codex import CodexService
from .config import BrokerConfig, load_config
from .session_store import AuthSessionStore


@dataclass
class BrokerRuntime:
    config: BrokerConfig
    auth_service: AuthService
    codex_service: CodexService


def create_runtime(config: BrokerConfig | None = None) -> BrokerRuntime:
    resolved = config or load_config()
    auth_store = AuthSessionStore(resolved.auth_store_path)
    auth_service = AuthService(session_store=auth_store)
    auth_service.initialize()

    codex_service = CodexService(
        auth_service=auth_service,
        base_url=resolved.codex_base_url,
        user_agent=resolved.user_agent,
    )

    return BrokerRuntime(
        config=resolved,
        auth_service=auth_service,
        codex_service=codex_service,
    )
